### PR TITLE
Cast efficiency rework

### PR DESCRIPTION
**Pull request by sMteX**
Okay, so what I tried to do:

1. In `Parser/Core/Modules/CastEfficiency.js` are now the most used spell categories (rotational, utility, others, items etc.)
2. I also added shared legendaries and trinkets/items to `CPM_ABILITIES`
3. Spec `CastEfficiency` modules were reworked, so that they use the `CPM_ABILITIES` from the Core module by adding `...CoreCastEfficiency.CPM_ABILITIES,` on the beginning of the `CPM_ABILITIES` (this way they don't have to worry about adding shared items/abilities)
4. If your spec added a specific category (I think Elemental Shaman has 'Dots'), you can do it by (for example) 
`static SPELL_CATEGORIES = {
    ...CoreCastEfficiency.SPELL_CATEGORIES,
    DOTS: 'Dots',
};`
5. Since some specs would now have empty categories (because they are baseline), I added a filter into `Main/CastEfficiency` component that filters out empty categories so only categories with at least one item in it should be shown.

I haven't changed anything else but things mentioned before in other specs so things should still work as they did before, but I still advise all spec maintainers to take a look at these changes and tell me if they disagree with anything.

Also, several specs weren't updated to the new CastEfficiency modules. In order not to break anything, I added those modules (and tested on some random logs, works just fine). 

EDIT: **Elemental Shaman** is still broken - the Cast Efficiency tab shows up (at least it doesn't crash now) but it needs some individual rework, the data it wants to show just isn't coming in and it messes up the whole table.

Also, unrelated to reworking CastEfficiency modules, I split the Archimonde's Hatred Reborn into its item part and the absorb part (listed now in `SPELLS_OTHERS.js`) and added that along the other trinkets to `CPM_ABILITIES`. Although it could be perhaps left alone, I think it would crash if added into the `CPM_ABILITIES`, or it would list the item's id (because that's what's under the `id`) and not the spell's id (which had `absorbId`). To compensate for these changes, I also updated the ArchimondesHatredReborn module as well. It all works just fine.

### CODE REVIEW

**Comment by fasib** [comment]
Found the Elemental Bug:

https://github.com/MartijnHols/WoWAnalyzer/blob/master/src/Parser/ElementalShaman/Modules/Features/CastEfficiencyComponent.js#L32
```
                    <SpellLink id={ability.spellId} style={{ color: '#fff' }}>
                      <SpellIcon id={ability.spellId} noLink /> {name}
                    </SpellLink>
```
should be changed to
```
                      <SpellLink id={ability.spell.id} style={{ color: '#fff' }}>
                        <SpellIcon id={ability.spell.id} noLink /> {ability.spell.name}
                      </SpellLink>
```

**Comment by sMteX** [comment]
I found the id but missed the name. I'll try on my local if that fixes it.

**Comment by sMteX** [comment]
Ok, I changed the name as well (few lines up `const name = ability.spell.name`), and at least it now doesn't crash, but the table is still wrong. From code I think it's supposed to show also Spend and Generated Maelstrom, but that's based on category keys and those are definitely different. Four additional columns aren't showing due to the different object structure.

It needs some kind of rework, but at least it doesn't crash now.

File: src/Parser/Core/Modules/CastEfficiency.js [hunk 1]
```diff
@@ -13,65 +13,80 @@ import SPELLS from 'common/SPELLS';
...
+    COOLDOWNS: 'Cooldown',
+    OTHERS: 'Spell',
+    UTILITY: 'Utility',
+    ITEMS: 'Item',
+  };
   static CPM_ABILITIES = [
     //Shared trinkets and legendaries
     {
       spell: SPELLS.CLEANSING_MATRIX,
+      name: `${ITEMS.ARCHIVE_OF_FAITH.name}`,
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/CastEfficiency.js]: Nice, we should do this for all items

File: src/Parser/Core/Modules/CastEfficiency.js [hunk 2]
```diff
@@ -13,65 +13,80 @@ import SPELLS from 'common/SPELLS';
...
     {
       spell: SPELLS.GUIDING_HAND,
+      name: `${ITEMS.DECEIVERS_GRAND_DESIGN.name}`,
       charges: 2,
+      category: CastEfficiency.SPELL_CATEGORIES.ITEMS,
       getCooldown: haste => 120,
       isActive: combatant => combatant.hasTrinket(ITEMS.DECEIVERS_GRAND_DESIGN.id),
     },
     {
       spell: SPELLS.GNAWED_THUMB_RING,
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/CastEfficiency.js]: such as this 😅

File: src/Parser/Core/Modules/CastEfficiency.js [hunk 3]
```diff
@@ -13,65 +13,80 @@ import SPELLS from 'common/SPELLS';
...
+      category: CastEfficiency.SPELL_CATEGORIES.ITEMS,
       getCooldown: haste => 180,
       isActive: combatant => combatant.hasFinger(ITEMS.GNAWED_THUMB_RING.id),
     },
     {
       spell: SPELLS.KILJAEDENS_BURNING_WISH_CAST,
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/CastEfficiency.js]: and this 😅

File: src/Parser/Core/Modules/CastEfficiency.js [hunk 4]
```diff
@@ -13,65 +13,80 @@ import SPELLS from 'common/SPELLS';
...
+      category: CastEfficiency.SPELL_CATEGORIES.ITEMS,
       getCooldown: haste => 75,
       isActive: combatant => combatant.hasTrinket(ITEMS.KILJAEDENS_BURNING_WISH.id),
     },
     {
+      spell: SPELLS.ARCHIMONDES_HATRED_REBORN_ABSORB,
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/CastEfficiency.js]: and this 😅

File: src/Parser/Core/Modules/CastEfficiency.js [hunk 5]
```diff
@@ -13,65 +13,80 @@ import SPELLS from 'common/SPELLS';
...
     },
     {
       spell: SPELLS.TOME_OF_UNRAVELING_SANITY_DAMAGE,
+      name: `${ITEMS.TOME_OF_UNRAVELING_SANITY.name}`,
+      category: CastEfficiency.SPELL_CATEGORIES.ITEMS,
       getCooldown: haste => 60,
       isActive: combatant => combatant.hasTrinket(ITEMS.TOME_OF_UNRAVELING_SANITY.id),
     },
     {
       spell: SPELLS.VELENS_FUTURE_SIGHT,
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/CastEfficiency.js]: and this 😅

File: src/Parser/ElementalShaman/Modules/Features/CastEfficiency.js [hunk 6]
```diff
@@ -10,92 +10,97 @@ import Tab from 'Main/Tab';
...
-      getCooldown: haste => 75,
-      isActive: combatant => combatant.hasTrinket(ITEMS.VELENS_FUTURE_SIGHT.id),
-    },
-    {
-      spell: SPELLS.GNAWED_THUMB_RING,
-      category: SPELL_CATEGORY.COOLDOWNS,
-      getCooldown: haste => 180,
-      isActive: combatant => combatant.hasFinger(ITEMS.GNAWED_THUMB_RING.id),
-    },
+    // {
```
Comment (MartijnHols) [inline src/Parser/ElementalShaman/Modules/Features/CastEfficiency.js]: Better to remove these then leave them behind as comments forever

File: src/Parser/ElementalShaman/Modules/Features/CastEfficiencyComponent.js [hunk 7]
```diff
@@ -24,33 +24,33 @@ const CastEfficiency = ({ categories, abilities }) => {
...
-                    {wasted ? wasted : ''}
-                  </td>
-                  <td style={{ width: '25%', color: 'orange' }}>
-                    {canBeImproved && !ability.noCanBeImproved && 'Can be improved.'}
-                  </td>
-                </tr>
-              );
-            })}
+              .map(({ ability, casts, spend, created, wasted, canBeImproved }) => {
+                const name = ability.spell.name;
```
Comment (MartijnHols) [inline src/Parser/ElementalShaman/Modules/Features/CastEfficiencyComponent.js]: should be `ability.name || ability.spell.name`

File: src/Parser/EnhancementShaman/Modules/Main/CastEfficiency.js [hunk 8]
```diff
@@ -98,8 +94,8 @@ class CastEfficiency extends CoreCastEfficiency {
       render: () => (
         <Tab title="Cast efficiency">
           <CastEfficiencyComponent
-            categories={this.constructor.SPELL_CATEGORIES}
-            abilities={getCastEfficiency(this.constructor.CPM_ABILITIES, this.owner)}
+            categories={CastEfficiency.SPELL_CATEGORIES}
```
Comment (MartijnHols) [inline src/Parser/EnhancementShaman/Modules/Main/CastEfficiency.js]: this.constructor is cleaner

File: src/Parser/RestoDruid/CombatLogParser.js [hunk 9]
```diff
@@ -10,11 +10,11 @@ import StatisticBox from 'Main/StatisticBox';
 import SuggestionsTab from 'Main/SuggestionsTab';
 import Tab from 'Main/Tab';
 import Talents from 'Main/Talents';
-import CastEfficiency from 'Main/CastEfficiency';
+// import CastEfficiency from 'Main/CastEfficiency';
```
Comment (MartijnHols) [inline src/Parser/RestoDruid/CombatLogParser.js]: this is better removed than left forever as comments

File: src/Parser/RestoDruid/CombatLogParser.js [hunk 10]
```diff
@@ -357,18 +359,18 @@ class CombatLogParser extends MainCombatLogParser {
...
-      if (cpm.canBeImproved && !cpm.ability.noSuggestion) {
-        results.addIssue({
-          issue: <span>Try to cast <SpellLink id={cpm.ability.spell.id} /> more often.</span>,
-          stat: `${cpm.casts} out of ${cpm.maxCasts} possible casts; ${Math.round(cpm.castEfficiency * 100)}% cast efficiency (>${cpm.recommendedCastEfficiency * 100}% is recommended)`,
-          icon: cpm.ability.spell.icon,
-          importance: cpm.ability.importance || getIssueImportance(cpm.castEfficiency, cpm.recommendedCastEfficiency - 0.05, cpm.recommendedCastEfficiency - 0.15),
-        });
-      }
-    });
+    // const castEfficiencyCategories = SPELL_CATEGORY;
```
Comment (MartijnHols) [inline src/Parser/RestoDruid/CombatLogParser.js]: this too

File: src/Parser/RestoDruid/CombatLogParser.js [hunk 11]
```diff
@@ -711,18 +713,18 @@ class CombatLogParser extends MainCombatLogParser {
...
-        render: () => (
-          <Tab title="Cast efficiency">
-            <CastEfficiency
-              categories={castEfficiencyCategories}
-              abilities={castEfficiency}
-            />
-          </Tab>
-        ),
-      },
+      // {
```
Comment (MartijnHols) [inline src/Parser/RestoDruid/CombatLogParser.js]: this too

File: src/Parser/RestorationShaman/CombatLogParser.js [hunk 12]
```diff
@@ -10,12 +10,12 @@ import StatisticBox from 'Main/StatisticBox';
 import SuggestionsTab from 'Main/SuggestionsTab';
 import Tab from 'Main/Tab';
 import Talents from 'Main/Talents';
-import CastEfficiency from 'Main/CastEfficiency';
+// import CastEfficiency from 'Main/CastEfficiency';
```
Comment (MartijnHols) [inline src/Parser/RestorationShaman/CombatLogParser.js]: more dead comments :P

**Comment by MartijnHols** [comment]
Thanks for the PR, looks good! 👍 Usually when you're commenting out code it's better to remove it. Git keeps it in history in case you ever needs it back, and otherwise it will forever be there. Please review the comments when you can.

**Comment by MartijnHols** [comment]
Also it's important the TravisCI check passes, otherwise your code can't be merged. Click the *details* link of the check and scroll down to find out what's blocking the build.

File: src/common/ITEMS.js [hunk 13]
```diff
@@ -36,7 +36,7 @@ const ITEMS = {
   },
   ARCHIMONDES_HATRED_REBORN: {
     id: 144249,
-    absorbId: 235169,
+    //absorbId: 235169,
```
Comment (MartijnHols) [inline src/common/ITEMS.js]: Should be removed instead of commented out, otherwise this will remain here *forever*

**Comment by MartijnHols** [comment]
Thanks!

File: src/common/ITEMS.js (hunk 13 above)
Comment (sMteX) [inline src/common/ITEMS.js]: Damn, I forgot about that one. Sorry, will fix in next PR :)

**Pull request merged**